lcls_tools/cor_plot/cor_plot_mat_scan.py

```diff
@@ -0,0 +1,243 @@
+
+    @property
+    def file(self):
+        """Loaded .mat file"""
+        return self._file
+
+    @property
+    def fields(self):
+        """Data fields (keys) Henrik felt like populating for given scan, depends
+        on what boxes are checked for measurements, so we'll never know unless
+        we introspect because schema is not a thing in Henrik's code.  I've never
+        seen code so uninterested in organizing data in an appreciable way...
+        probably for job security?"""
+        return self._fields
+
+    @property
+    def control_dict(self):
+        """Ctrl PV and vals, units etc..."""
+        return self._ctrl_dict
+
+    @property
+    def accelerator(self):
+        """Accelerator name"""
+        return self._accel
+
+    @property
+    def ctrl_pv(self):
+        """PV being scanned, might need to change for 2D scans"""
+        if self._ctrl_dict is not None:
+            return self._ctrl_dict[0]['name']
+
+        return None
+
+    @property
+    def iterations(self):
+        """iterations for cor plot"""
+        if self._ctrl_dict is not None:
+            return len(self._ctrl_dict[0]['vals'])
+
+        return 0
+
+    @property
+    def ctrl_vals(self):
+        """Vals for ctrl pv in scan"""
+        if self._ctrl_dict is not None:
+            return self._ctrl_dict[0]['vals']
+
+        return None
+
+    @property
+    def beam(self):
+        """This is a huge amount of data"""
+        return self._beam
+
+    @property
+    def beam_names(self):
+        """The different keys for a beam dict for each iteration and fit"""
+        return self._beam_names
+
+    @property
+    def timestamp(self):
+        """Matlab timestamp (ordinal)"""
+        return self._ts
+
+    @property
+    def config(self):
+        """Random collection of data already available from above properties, lame"""
+        return self._config
+
+    @property
+    def samples(self):
+        if self._beam:
+            return len(self._beam[0])
+
+    def _unpack_accl(self, data):
+        """Accelerator name such as LCLS, LCLS2, etc..."""
+        # Need to write generic decorator for this behavior
+        if ACCL not in self._fields:
+            return None
+
+        idx = self._fields.index(ACCL)
+        return str(data[idx][0])
+
+    def _unpack_statuses(self, data):
+        """Pull statuses out of the nested array"""
+        if STAT not in self._fields:
+            return None
+
+        idx = self._fields.index(STAT)
+        statuses = data[idx]
+        return [status[0] for status in statuses]
+
+    def _unpack_ctrl_pv(self, data):
+        """Produces a list of ctrl pv dicts with list of vals and 
+        tmestamps.
+        """
+        if CTRL not in self._fields:
+            return None
+
+        idx = self._fields.index(CTRL)
+        ctrls = data[idx]
+        ctrl_list = []
+        for ctrl in ctrls:
+            temp = dict()
+            temp['name'] = str(ctrl[0][0][0])
```

Comment:
yikes on this triple 0 index. Just putting a pin in this for later, not sure if there's anything else we can do....

Reply:
Yeah, Matlab structures come through as giant nested lists (if only he had used Matlab dictionaries).  What could be done is if someone revamped the Matlab data saving code, but it's fairly hideous... poor Tim